libzecale/circuits/pairing/bls12_377_pairing.hpp

```diff
@@ -325,6 +325,119 @@ class bls12_377_ate_miller_loop_gadget : public libsnark::gadget<libff::Fr<ppT>>
     void generate_r1cs_witness();
 };
 
+template<typename ppT>
+class bls12_377_final_exp_first_part_gadget
+    : public libsnark::gadget<libff::Fr<ppT>>
+{
+public:
+    using FieldT = libff::Fr<ppT>;
+    using FqkT = libff::Fqk<libsnark::other_curve<ppT>>;
+
+    // Follows the implementation used in
+    // libff::bls12_377_final_exponentiation_first_chunk() (see
+    // libff/algebra/curves/bls12_377/bls12_377_pairing.cpp), which in turn
```

Comment:
`libff/algebra/curves/bls12_377/bls12_377_pairing.cpp` -> `clearmatics/libff/algebra/curves/bls12_377/bls12_377_pairing.cpp` (the scipr-lab `libff` library does not yet contain this code)

Reply:
Note: This is applicable to other places in this PR

Reply:
changed references to clearmatics/libff

Reply:
Just as a quick note, remember that `libff` contains a `libff` repo that contains the source of the library. Strictly speaking, we need `clearmatics/libff/libff`, but that's a detail

Reply:
Good point.  I'll fix this as part of #19